src/utilities-selectors.js

```diff
@@ -0,0 +1,245 @@
+import {
+  convertMatchListToRegExp,
+  flattenArray,
+  getCombinations,
+} from './utilities-data';
+import {testSelector} from './utilities-dom';
+import {getTagSelector} from './selector-tag';
+import {getIdSelector} from './selector-id';
+import {getClassSelectors} from './selector-class';
+import {getAttributeSelectors} from './selector-attribute';
+
+export const ESCAPED_COLON = ':'
+  .charCodeAt(0)
+  .toString(16)
+  .toUpperCase();
+
+// Square brackets need to be escaped, but eslint has a problem with that.
+/* eslint-disable-next-line no-useless-escape */
+export const SPECIAL_CHARACTERS_RE = /[ !"#$%&'()\[\]{|}<>*+,./;=?@^`~\\]/;
+
+/**
+ * Escapes special characters used by CSS selector items.
+ * @param {string} input
+ * @return {string}
+ */
+export function sanitizeSelectorItem (input = '') {
+  return input.split('')
+    .map((character) => {
+      if (character === ':') {
+        return `\\${ESCAPED_COLON} `;
+      }
+      if (SPECIAL_CHARACTERS_RE.test(character)) {
+        return `\\${character}`;
+      }
+      return escape(character)
+        .replace(/%/g, '\\');
+    })
+    .join('');
+}
+
+export const SELECTOR_TYPE_GETTERS = {
+  tag: getTagSelector,
+  id: getIdSelector,
+  class: getClassSelectors,
+  attribute: getAttributeSelectors,
+  nthchild: getNthChildSelector,
+};
+
+/**
+ * Returns list of selectors of given type for the element.
+ * @param {Element} element
+ * @param {string} selector_type
+ * @return {Array.<string>} - Always an array, even if the selector only allows single value (e.g. tag).
+ */
+export function getSelectorsByType (element, selector_type) {
+  return (SELECTOR_TYPE_GETTERS[selector_type] || (() => []))(element);
+}
+
+/**
+ * Remove blacklisted selectors from list.
+ * @param {Array.<string>} list
+ * @param {RegExp} blacklist_re
+ * @return {Array.<string>}
+ */
+export function filterSelectors (list = [], blacklist_re) {
+  return list.filter((item) => !blacklist_re.test(item));
+}
+
+/**
+ * Prioritise whitelisted selectors in list.
+ * @param {Array.<string>} list
+ * @param {RegExp} whitelist_re
+ * @return {Array.<string>}
+ */
+export function orderSelectors (list = [], whitelist_re) {
+  return list.sort((a, b) => {
+    const a_is_whitelisted = whitelist_re.test(a);
+    const b_is_whitelisted = whitelist_re.test(b);
+    if (a_is_whitelisted && !b_is_whitelisted) {return -1;}
+    if (!a_is_whitelisted && b_is_whitelisted) {return 1;}
+    return 0;
+  });
+}
+
+/**
+ * Tries to generate unique selector for the element within it's parent.
+ * @param {Element} element
+ * @param {css_selector_generator_options} options
+ * @return {string} Either unique selector or "*" if not possible.
+ */
+export function getUniqueSelectorWithinParent (element, options) {
+  const selectors_list = getSelectorsList(element, options);
+  const type_combinations = getTypeCombinations(selectors_list, options);
+  const all_selectors = flattenArray(type_combinations);
```

Comment:
As I understand, this implies generating all possible generators in order to test their uniqueness in the for-loop below.

I wonder how expensive this operation is and whether a way to optimize this with few changes wouldn't be the use of [Generators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator).

The main advantage of generators is that it returns an iterator that you can browse with a classical `for-of` loop.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Iterators_and_Generators

Reply:
Yeah, I was thinking about using them. But I got stuck on something... I think it was about not being able to generate the candidates in the right order. Sorting the array was easier.

I think it would be nice playground for performance optimisation later.

Reply:
Sounds wise :)